pkg/network/netlink/circuit_breaker.go

```diff
@@ -0,0 +1,126 @@
+package netlink
+
+import (
+	"math"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	tickInterval  = 3 * time.Second
+	breakerOpen   = int64(1)
+	breakerClosed = int64(0)
+
+	// The lower this number is the more amortized the average is
+	// For example, if ewmaWeight is 1, a single burst of events might
+	// cause the breaker to trip.
+	ewmaWeight = 0.2
+)
+
+// CircuitBreaker is meant to enforce a maximum rate of events per second
+// Once the event rate goes above the threshold the circuit breaker will trip
+// and remain open until Reset() is called.
+type CircuitBreaker struct {
+	// The maximum rate of events allowed to pass
+	maxEventsPerSec int64
+
+	// The number of events elapsed since the last tick
+	eventCount int64
+
+	// An exponentially weighted average of the event rate (per second)
+	// This is what actually compare against maxEventsPersec
+	eventRate int64
+
+	// Represents the status of the cicuit breaker.
+	// 1 means open, 0 means closed
+	status int64
+
+	// The timestamp in nanoseconds of when we last updated eventRate
+	lastUpdate int64
+}
+
+// NewCircuitBreaker instantiates a new CircuitBreaker that only allows
+// a maxEventsPerSec to pass. The rate of events is calculated using an EWMA.
+func NewCircuitBreaker(maxEventsPerSec int) *CircuitBreaker {
+	// -1 will virtually disable the circuit breaker
+	if maxEventsPerSec == -1 {
+		maxEventsPerSec = math.MaxInt64
+	}
+
+	c := &CircuitBreaker{maxEventsPerSec: int64(maxEventsPerSec)}
+	c.Reset()
+
+	go func() {
+		ticker := time.NewTicker(tickInterval)
+		for t := range ticker.C {
+			c.update(t)
+		}
+	}()
+
+	return c
+}
+
+// IsOpen returns true when the circuit breaker trips and remain
+// unchanched until Reset() is called.
+func (c *CircuitBreaker) IsOpen() bool {
+	return atomic.LoadInt64(&c.status) == breakerOpen
+}
+
+// Tick represents one or more events passing through the circuit breaker.
+func (c *CircuitBreaker) Tick(n int) {
+	atomic.AddInt64(&c.eventCount, int64(n))
+}
+
+// Rate returns the current rate of events
+func (c *CircuitBreaker) Rate() int64 {
+	return atomic.LoadInt64(&c.eventRate)
+}
+
+// Reset closes the circuit breaker and its state.
+func (c *CircuitBreaker) Reset() {
+	atomic.StoreInt64(&c.eventCount, 0)
+	atomic.StoreInt64(&c.status, breakerClosed)
+	atomic.StoreInt64(&c.eventRate, 0)
+	atomic.StoreInt64(&c.lastUpdate, time.Now().UnixNano())
+}
```

Comment:
That's a good point.

We don't run into any risks here though, because `Reset()` is only called after the circuit breaker trips, and when the circuit breaker is open, there is an early return in the `update()` method.

Edit: actually there is a small code change I should do add the end of `update()`:

```
	if int64(newEventRate) > c.maxEventsPerSec {
		atomic.StoreInt64(&c.status, breakerOpen)
	}

	atomic.StoreInt64(&c.lastUpdate, now.UnixNano())
	atomic.StoreInt64(&c.eventRate, int64(newEventRate))
```

should be changed to:

```
	atomic.StoreInt64(&c.lastUpdate, now.UnixNano())
	atomic.StoreInt64(&c.eventRate, int64(newEventRate))

	if int64(newEventRate) > c.maxEventsPerSec {
		atomic.StoreInt64(&c.status, breakerOpen)
	}
```

to avoid any chance of data race.

Reply:
This is done.

Reply:
I still think this might be problematic.

Reset looks like

```func (c *CircuitBreaker) Reset() {
	atomic.StoreInt64(&c.eventCount, 0)
	atomic.StoreInt64(&c.status, breakerClosed)
	atomic.StoreInt64(&c.eventRate, 0)
	atomic.StoreInt64(&c.lastUpdate, time.Now().UnixNano())
}
```

If the following two statements were to execute:

```
	atomic.StoreInt64(&c.eventCount, 0)
	atomic.StoreInt64(&c.status, breakerClosed)
```

And then the goroutine was paused, `update` could then start executing.  The breaker is now closed so it will proceed through.

It may get down to executing the following block

```
        eventCount := atomic.SwapInt64(&c.eventCount, 0)
	prevEventRate := atomic.LoadInt64(&c.eventRate)
	newEventRate := ewmaWeight*float64(eventCount)/deltaInSec + (1-ewmaWeight)*float64(prevEventRate)
```

`eventCount` will have already been reset but `eventRate` will not.

What effect would that have?

Reply:
oh, I think you're right. 
In that case we can simply move  `atomic.StoreInt64(&c.status, breakerClosed)` to the bottom of  the `Reset()` method, no?

Reply:
The breaker could still be closed with valid data in eventCount/eventRate right (if it is functioning normally and not tripping) ?

Reply:
oh, so `Reset()` is only called *after* the circuit breaker trips. I agree that hidden contract is not great in terms of API but given that it's only used inside `Consumer` it should be fine?
I could add comment to `Reset()` clarifying that if you think it's worth it.